procrastinate/worker.py

```diff
@@ -152,25 +152,26 @@ async def process_job(self, job: jobs.Job, worker_id: int = 0) -> None:
             extra=context.log_extra(action="loaded_job_info"),
         )
 
-        status = jobs.Status.FAILED
-        next_attempt_scheduled_at = None
+        status, retry_at = None, None
         try:
             await self.run_job(job=job, worker_id=worker_id)
             status = jobs.Status.SUCCEEDED
         except exceptions.JobRetry as e:
-            status = jobs.Status.TODO
-            next_attempt_scheduled_at = e.scheduled_at
+            retry_at = e.scheduled_at
         except exceptions.JobError:
-            pass
+            status = jobs.Status.FAILED
         except exceptions.TaskNotFound as exc:
+            status = jobs.Status.FAILED
             self.logger.exception(
                 f"Task was not found: {exc}",
                 extra=context.log_extra(action="task_not_found", exception=str(exc)),
             )
         finally:
-            await self.job_manager.finish_job(
-                job=job, status=status, scheduled_at=next_attempt_scheduled_at
-            )
+            if retry_at:
```

Comment:
juste une question svp :)  : est-ce que cette gestion faite en python pourrait pas être faite plutôt dans le schema ?
et même que la procédure stockée en db fasse elle même un pg cron pour faire un notify à la date voulue ?

Reply:
Hey :) This is a public project, I know you know I speak French, but it's easier for other people if we try and stick to English ;)

As of today, determining if a task should be retried or not is done in Python, because it is an operation that depends on the Python outcome of running the task. Could you tell us more on what you're suggesting ? Also, I'm not really sure it belongs to this PR, but maybe I misunderstood your comment.

Reply:
Hello. yes sorry ;) just wondering if the logic to retry of not could be handled in the 
_procrastinate_finish_job_   stored procedure. So that, depending if _next_scheduled_at_ is null or not , the procedure would call _procrastinate_finish_job_1_ or _procrastinate_retry_job_ .
Moreother, inside the _procrastinate_retry_job_, we could imagine that a notify (call to procrastinate_notify_queue()) could be done at the right _next_scheduled_at_  thanks to a pg cron ?

Reply:
> just wondering if the logic to retry of not could be handled in the procrastinate_finish_job stored procedure. So that, depending if next_scheduled_at is null or not , the procedure would call procrastinate_finish_job_1 or procrastinate_retry_job

Well, that's basically how it's implemented currently (before this PR). Except that the the status should always be passed, even when `scheduled_at` is specified. See https://github.com/peopledoc/procrastinate/blob/ddd2d19794227b4178084d59f56f3f38f09873c8/procrastinate/sql/schema.sql#L175-L185.

And we thought that it would make for a cleaner and more explicit API if we had two separate functions. We want to distinguish the action of "finishing a job", which involves changing its status to either "succeeded" or "failed", from the action of "retrying a job", which involves resetting its status to "todo", with a scheduling in the future.

> Moreother, inside the _procrastinate_retry_job_, we could imagine that a notify (call to procrastinate_notify_queue()) could be done at the right _next_scheduled_at_ thanks to a pg cron ?

Are you referring to [`pg_cron`](https://github.com/citusdata/pg_cron)?

Reply:
Hello. thanks for your answer. yes, i'm refering to pg_cron.
What i m' suggesting is something like : 

```
CREATE FUNCTION procrastinate_finish_job (job_id integer, end_status procrastinate_job_status, next_scheduled_at timestamp with time zone) RETURNS void
BEGIN
   IF (next_scheduled_at IS NULL) 
         CALL procrastinate_finish_job_1(job_id, end_status)
  ELSE
         CALL procrastinate_retry_job(job_id, next_scheduled_at)
END
```

so that to minimize the code that the client has to do. The client only has to call _procrastinate_finish_job_ 
Remark : if guess we want to retry only on on failed status ?
Remark2: calling only the retry procedure with status='todo', you won't have anymore the _procrastinate_events_ failed status triggered by [ref](https://github.com/peopledoc/procrastinate/blob/ddd2d19794227b4178084d59f56f3f38f09873c8/procrastinate/sql/schema.sql#L260)